### PR TITLE
Onlinetraining

### PR DESCRIPTION
**Pull request by Error323**
This is still a work in progress, but this way reviews can be performed already. It's currently over 4 times slower (~1700 pos/s) than the binary implementation in `supervised_parse.py` (7000 pos/s). So 4 hours of training on the binary version takes 16 hours here...

Next up is switching internal representation to pure binary as leelazero did. This should bump the speed back up to ~7000 pos/s.

Ideally I would like to get to a version that doesn't create new ChunkParser objects and a new TFProcess object. But I'm not sure how yet.

### CODE REVIEW

**Comment by ywrt** [comment]
Hum. Yes, use a FileDataSrc the answer to the question you asked.

As an aside: I would recommend starting from https://github.com/gcp/leela-zero/pull/1003 and pulling out all the constants in chunkparser.py into a config block. That should let you easily re-use all that code.

The Alpha Zero paper uses the format `N×N×(MT+L)`, so defining 'N', M', 'T' and 'L' should be sufficient to compute all the various constants.

**Comment by Error323** [comment]
Currently maxing out at ~5400 pos/s using 10 cpus (inc hyperthreading). @ywrt it's not yet maxed out, but a _lot_ better already. I was thinking, since applying a symmetry occurs after the shufflebuffer extraction, wouldn't it make sense to allow multiple hits in the shufflebuffer before removal?

File: training/tf/parse.py [hunk 1]
```diff
@@ -30,203 +30,90 @@
...
+        self.chunks = chunks
+        random.shuffle(self.chunks)
+        self.done = []
+    def next(self):
+        if not self.chunks:
+            self.chunks = self.done
+            random.shuffle(self.chunks)
+        if not self.chunks:
+            return None
+        while (True):
```
Comment (ywrt) [inline training/tf/parse.py]: `while len(self.chunks):` would be better to avoid the potential infinite loop.

File: training/tf/parse.py [hunk 2]
```diff
@@ -30,203 +30,90 @@
 import time
 import tensorflow as tf
 from tfprocess import TFProcess
+from chunkparser import ChunkParser
 
-DATA_ITEM_LINES = 121
-SKIP = 16
-CHUNK_PASSES = 2
+SKIP = 8
+RAM_BATCH_SIZE = 2048
```
Comment (ywrt) [inline training/tf/parse.py]: Does this really fit in the GPU RAM!?

Note that the memory for the batch itself is ~nothing: The common problem is that TensorFlow by default accumulates all the gradients from each example separately and then sums them up at the end.

So this effectively results in 2048 copies of the network in GPU RAM, while probably only works for either very small networks or uncommonly huge hardware.

**Comment by ywrt** [comment]
Re sampling multiple symmetries in the same batch: I'd be too worried about contaminating training. mini-batch SGD is *really* good at finding patterns in the input that you didn't want it to find ;)

And yes, the V2 format will easily max out the GPU. It does 25,000 pos/s on my (old) machine using 2 workers, so it should be able to do substantially better than that for you.

File: training/tf/parse.py (hunk 2 above)
Comment (Error323) [reply training/tf/parse.py]: Oh darn. You're right that makes no sense. I've got 2 1080Ti, but this uses 1. What would be a sane value?

Comment (Error323) [reply training/tf/parse.py]: The current net has 4'214'045 weights, so about 16M. I guess RAM_BATCH_SIZE=512 should work?

**Comment by Error323** [comment]
It won't be in the same batch persee, but in the same shufflebuffer. I'm suggesting to allow e.g. 2 hits from an sb index before removal. The size of the shufflebuffer would still dictate the probability of 2 symmetries ending up in the same batch.

**Comment by ywrt** [comment]
I wouldn't see it as worth it if you're using the V2 format. I use v2_write_training.py to generate pre-shuffled data, which lets me set SKIP=1, which results in ~ no cpu for the data feed pipeline.
Which means there's no point in re-sampling the shuffle buffer.

File: training/tf/parse.py (hunk 2 above)
Comment (Error323) [reply training/tf/parse.py]: It seems a RAM_BATCH_SIZE of 512 results in only 1.3G gpu usage with tensorflow? Do you know how that's possible?

**Comment by Error323** [comment]
That makes sense, we should just switch to V2 asap. Another question: Does every worker get a deep copy of FileDataSrc? Because they all seem to be hitting the same faulty chunk at about the same time (I have 2 faulty chunks in the dataset).

**Comment by ywrt** [comment]
They all get a pickled copy. And yes, you're correct that there's a bug there: the shuffle should be done outside the init. I'll fix.

**Comment by ywrt** [comment]
NB: fix is trivial. Just change `__init__` to:
```
class FileDataSrc: 
  def __init__(self, chunks):
       self.done = chunks
       self.chunks = []
```
I wrote the next() code to support that but forgot to change the init code. doh! :)

**Comment by Error323** [comment]
@ywrt was training already performed with this code? It doesn't seem to converge for me and the nof weights seems to differ from our original implementation.

**Comment by jkiliani** [comment]
Have you tried training a Go neural network on LZ data with this code, e.g. with some of the training data @gcp puts online regularly?

**Comment by Error323** [comment]
@jkiliani, no I haven't still running experiments which are sucking up all my compute.

It must be some problem with the data being fed into tensorflow training. The trainingdata takes a nosedive on MSE:

```
step 100, policy=6.17128 mse=0.174783 reg=0.373261 total=7.24368 (4083.19 pos/s)
step 200, policy=5.18683 mse=0.038592 reg=0.374614 total=5.71581 (4920.47 pos/s)
step 300, policy=4.46734 mse=0.0172791 reg=0.374793 total=4.91125 (5090.22 pos/s)
step 400, policy=4.1405 mse=0.019355 reg=0.375308 total=4.59322 (5200.62 pos/s)
step 500, policy=4.16227 mse=0.0224253 reg=0.37795 total=4.62992 (5184.79 pos/s)
step 600, policy=3.6538 mse=0.00873278 reg=0.37824 total=4.06697 (4912.91 pos/s)
step 700, policy=3.44481 mse=0.00760179 reg=0.378191 total=3.85341 (4765.92 pos/s)
step 800, policy=3.27602 mse=0.0066682 reg=0.378092 total=3.68079 (4689.62 pos/s)
step 900, policy=3.16693 mse=0.00602719 reg=0.377943 total=3.56898 (4955.34 pos/s)
step 1000, policy=3.07752 mse=0.00538145 reg=0.377756 total=3.4768 (4822.89 pos/s)
step 1000, policy=7.69281 training accuracy=1.13037%, mse=0.465906
Model saved in file: /home/fhuizing/Workspace/chess/leela-chess/training/tf/leelaz-model-1000
Leela weights saved to /home/fhuizing/Workspace/chess/leela-chess/training/tf/leelaz-model-1000.txt
step 1100, policy=2.98752 mse=0.00480509 reg=0.3775 total=3.38424 (4037.93 pos/s)
step 1200, policy=2.91516 mse=0.00459634 reg=0.377179 total=3.31072 (4928.08 pos/s)
step 1300, policy=2.86306 mse=0.00449471 reg=0.376797 total=3.25784 (5180.59 pos/s)
```

**Comment by jkiliani** [comment]
It seems you're not the only one having problems with the LZ V2 pipeline: https://github.com/gcp/leela-zero/issues/1025
